src/backend/distributed/commands/local_multi_copy.c

```diff
@@ -0,0 +1,203 @@
+/*-------------------------------------------------------------------------
+ *
+ * local_multi_copy.c
+ *    Commands for running a copy locally
+ *
+ * For each local placement, we have a buffer. When we receive a slot
+ * from a copy, the slot will be put to the corresponding buffer based
+ * on the shard id. When the buffer size exceeds the threshold a local
+ * copy will be done. Also If we reach to the end of copy, we will send
+ * the current buffer for local copy.
+ *
+ * The existing logic from multi_copy.c and format are used, therefore
+ * even if user did not do a copy with binary format, it is possible that
+ * we are going to be using binary format internally.
+ *
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+#include "commands/copy.h"
+#include "catalog/namespace.h"
+#include "parser/parse_relation.h"
+#include "utils/lsyscache.h"
+#include "nodes/makefuncs.h"
+#include "safe_lib.h"
+#include <netinet/in.h> /* for htons */
+
+#include "distributed/transmit.h"
+#include "distributed/commands/multi_copy.h"
+#include "distributed/multi_partitioning_utils.h"
+#include "distributed/local_executor.h"
+#include "distributed/local_multi_copy.h"
+#include "distributed/shard_utils.h"
+
+/*
+ * LOCAL_COPY_BUFFER_SIZE is buffer size for local copy.
+ * There will be one buffer for each local placement, therefore
+ * the maximum amount of memory that might be alocated is
+ * LOCAL_COPY_BUFFER_SIZE * #local_placement
+ */
+#define LOCAL_COPY_BUFFER_SIZE (1 * 512 * 1024)
+
+
+static int ReadFromLocalBufferCallback(void *outbuf, int minread, int maxread);
+static void AddSlotToBuffer(TupleTableSlot *slot, CitusCopyDestReceiver *copyDest,
+							bool isBinary);
+
+static bool ShouldSendCopyNow(StringInfo buffer);
+static void DoLocalCopy(StringInfo buffer, Oid relationId, int64 shardId,
+						CopyStmt *copyStatement, bool isEndOfCopy);
+static bool ShouldAddBinaryHeaders(StringInfo buffer, bool isBinary);
+
+/*
+ * localCopyBuffer is used in copy callback to return the copied rows.
+ * The reason this is a global variable is that we cannot pass an additional
+ * argument to the copy callback.
+ */
+StringInfo localCopyBuffer;
+
+/*
+ * ProcessLocalCopy adds the given slot and does a local copy if
+ * this is the end of copy, or the buffer size exceeds the threshold.
+ */
+void
+ProcessLocalCopy(TupleTableSlot *slot, CitusCopyDestReceiver *copyDest, int64 shardId,
+				 StringInfo buffer, bool isEndOfCopy)
+{
+	/*
+	 * Here we save the previous buffer, and put the local shard's buffer
+	 * into copyOutState. The motivation is to use the existing logic to
+	 * serialize a row slot into buffer.
+	 */
+	StringInfo previousBuffer = copyDest->copyOutState->fe_msgbuf;
+	copyDest->copyOutState->fe_msgbuf = buffer;
```

Comment:
I get why we're doing this, but it seems messy. Are there any alternatives?

Reply:
Initially I started modifying the serialization methods to take a buffer instead of CopyOutState but while serializing we also use some internal fields like delimiter etc so the refactoring is not a small one.

We could maybe clone the copyOutState but then we probably dont want to have #shards clones. And if we clone it once, we will need to do the same switching for the buffer.

Reply:
Creating one per shard might be ok. It's not a huge amount of memory. You could then use the fe_msgbuf instead of a separate buffer.

Reply:
I changed the code to use cloned copyOutState. So instead of copyDest->copyOutState we are using it now. Removed the separate buffer.